src/Microsoft.OData.Core/ObjectPool.cs

```diff
@@ -0,0 +1,91 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ObjectPool.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// A pool of objects.
+    /// </summary>
+    /// <typeparam name="T">The type of objects to pool.</typeparam>
+    /// <remarks>This object pool is not thread-safe and not intended to be shared across concurrent requests.
+    /// Consider using the DefaultObjectPool in 8.x once we drop support for netstandard 1.1
+    /// </remarks>
+    internal class ObjectPool<T>
+    {
+        private readonly Func<T> objectGenerator;
+        private readonly ObjectWrapper[] items;
+        private T firstItem;
+
+        /// <summary>
+        /// To initialize the object pool.
+        /// </summary>
+        /// <param name="objectGenerator">Used to create an instance of a <typeparamref name="T"/>.</param>
+        public ObjectPool(Func<T> objectGenerator)
+        {
+            this.objectGenerator = objectGenerator ?? throw new ArgumentNullException(nameof(objectGenerator));
+            items = new ObjectWrapper[32];
+        }
+
+        /// <summary>
+        /// Gets an object from the pool if one is available, otherwise creates one.
+        /// </summary>
+        /// <returns>A <typeparamref name="T"/>.</returns>
+        public T Get()
+        {
+            T item = firstItem;
+
+            if (item != null)
+            {
+                firstItem = default(T);
+                return item;
+            }
+
+            for (int i = 0; i < items.Length; i++)

```

Comment:
I think there's a potential micro-optimization here. You could treat this is a stack and keep track of the last index where an object was found. After retrieving an object, you decrement that index, it's likely that the decremented index will contain an object. If you want to return an object, the index after the decrement index will likely be empty. But for a small fix-sized array, this is probably not worth it.